compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -236,137 +142,640 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-            // the arguments hence we only do our usual type checking with
-            // the arguments who's types we do know.
-            let t = if c_variadic {
-                expected_arg_count
-            } else if tuple_arguments == TupleArguments {
-                args.len()
+        for idx in deferred_arguments {
+            let (compatible, _, _) = demand_compatible(idx, idx);
+            // Note that, unlike the first pass, we ignore the checked/coerced types,
+            // since we don't plan on running select_obligations_where_possible again
+            if compatible {
+                compatibility_matrix[idx][idx] = true;
             } else {
-                supplied_arg_count
+                call_appears_satisfied = false;
+            }
+        }
+
+        // If something above didn't typecheck, we've fallen off the happy path
+        // and we should make some effort to provide better error messages
+        if !call_appears_satisfied {
+            // The algorithm here is inspired by levenshtein distance and longest common subsequence.
+            // We'll try to detect 4 different types of mistakes:
+            // - An extra parameter has been provided that doesn't satisfy *any* of the other inputs
+            // - An input is missing, which isn't satisfied by *any* of the other arguments
+            // - Some number of arguments have been provided in the wrong order
+            // - A type is straight up invalid
+
+            // First, fill in the rest of our compatibility matrix
+            for i in 0..provided_arg_count {
+                for j in 0..minimum_input_count {
+                    if i == j {
+                        continue;
+                    }
+                    compatibility_matrix[i][j] = check_compatible(i, j);
+                }
+            }
+
+            // Obviously, detecting exact user intention is impossible, so the goal here is to
+            // come up with as likely of a story as we can to be helpful.
+            //
+            // We'll iteratively removed "satisfied" input/argument paris,
+            // then check for the cases above, until we've eliminated the entire grid
+            //
+            // We'll want to know which arguments and inputs these rows and columns correspond to
+            // even after we delete them, so these lookups will keep track of that
+            let mut input_indexes: Vec<usize> = (0..minimum_input_count).collect();
+            let mut arg_indexes: Vec<usize> = (0..provided_arg_count).collect();
+
+            // First, set up some utility functions for the algorithm below
+            // Remove a given input or argument from consideration
+            let eliminate_input = |mat: &mut Vec<Vec<bool>>, ii: &mut Vec<usize>, idx| {
+                if idx >= ii.len() {
+                    return; // FIXME: Should this ICE as a compiler bug?
+                }
```

Comment:
So, `eliminate_input` is a helper method to delete a column from the "compatibility matrix" which tracks which arguments could be passed to which parameters.  However, as we eliminate those columns, we still need to know which columns correspond to which original parameters.

For example, suppose I have `fn myFunc(a: A, b: B, c: C)`, and you call it as `fn myFunc(B {}, A {}, C {})`.

The compatibility matrix would be (only T's are shown for clarity)
```
[., T, .]
[T, ., .]
[., ., T]
```

and the `input_indexes` collection (ii) would be `[1,2,3]`.

After we detect that a and b are swapped, we eliminate that row and column, and end up with
```
[T]
```

but to know that that single column corresponds to the third input, `input_indexes` is `[3]`.

I can't remember why, but at some point we were eliminating columns past the end of `input_indexes`, and I through this in to unblock me while I finished the rest of the algorithm.  I'll investigate whether it's still needed.

Reply:
This didn't cause any problems on my initial test cases, so I've removed it; I still haven't run the full test suite to bless things, so this might need to be added back later.